cmd/agent/common/autodiscovery.go

```diff
@@ -0,0 +1,116 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-2020 Datadog, Inc.
+
+package common
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/providers"
+	"github.com/DataDog/datadog-agent/pkg/autodiscovery/scheduler"
+	"github.com/DataDog/datadog-agent/pkg/config"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+func setupAutoDiscovery(confSearchPaths []string, metaScheduler *scheduler.MetaScheduler) *autodiscovery.AutoConfig {
+	ad := autodiscovery.NewAutoConfig(metaScheduler)
+	ad.AddConfigProvider(providers.NewFileConfigProvider(confSearchPaths), false, 0)
+
+	// Autodiscovery cannot easily use config.RegisterOverrideFunc() due to Unmarshalling
+	var discoveredProviders []config.ConfigurationProviders
+	var discoveredListeners []config.Listeners
+	if config.Datadog.GetBool("autoconf_from_environment") {
+		discoveredProviders, discoveredListeners = discoverAutodiscoveryComponents()
+	}
+
+	// Register additional configuration providers
+	var CP []config.ConfigurationProviders
+	err := config.Datadog.UnmarshalKey("config_providers", &CP)
+	if err == nil {
+		// Add extra config providers
+		for _, name := range config.Datadog.GetStringSlice("extra_config_providers") {
+			CP = append(CP, config.ConfigurationProviders{Name: name, Polling: true})
+		}
+
+		for _, provider := range discoveredProviders {
+			CP = append(CP, provider)
+		}
+	} else {
+		log.Errorf("Error while reading 'config_providers' settings: %v", err)
+	}
+
```

Comment:
should we do this only when `config.Datadog.GetBool("autoconf_from_environment") == false` ? (to avoid activating the same listeners/providers multiple times. otherwise we can dedupe `CP`

Reply:
Normally the AD code already does dedupe later on, but I can double check.

Reply:
Added code to not add if already there as code that dedupes later on generates a warning (which could be LGTM warning)